src/osp/tasks/top_utils.h

```diff
@@ -0,0 +1,183 @@
+T& top_assign(ArrayView<entt::any> const topData, TopDataId id, ANY_T&& any)
+{
+    entt::any &rData = topData[std::size_t(id)];
+    rData = std::forward<ANY_T>(any);
+    return entt::any_cast<T&>(rData);
+}
+
+
+template <typename T>
+[[nodiscard]] T& top_get(ArrayView<entt::any> const topData, TopDataId const id)
+{
+    return entt::any_cast<T&>(topData[id]);
+}
+
+
+
+template<typename FUNCTOR_T>
+struct wrap_args_trait
+{
+    template<typename ... ARGS_T, std::size_t ... INDEX_T>
+    static constexpr void apply_void(ArrayView<entt::any> topData, [[maybe_unused]] std::integer_sequence<std::size_t, INDEX_T...> indices) noexcept
+    {
+        FUNCTOR_T{}(entt::any_cast<ARGS_T&>(topData[INDEX_T]) ...);
+    }
+
+    template<typename ... ARGS_T>
+    static TopTaskStatus wrapper_void([[maybe_unused]] WorkerContext& rCtx, ArrayView<entt::any> topData) noexcept
+    {
+        assert(topData.size() == sizeof...(ARGS_T));
+
+        apply_void<ARGS_T ...>(topData, std::make_integer_sequence<std::size_t, sizeof...(ARGS_T)>{});
+
+        return TopTaskStatus::Success;
+    }
+
+    template<typename ... ARGS_T, std::size_t ... INDEX_T>
+    static constexpr TopTaskStatus apply(ArrayView<entt::any> topData, [[maybe_unused]] std::integer_sequence<std::size_t, INDEX_T...> indices) noexcept
+    {
+        return FUNCTOR_T{}(entt::any_cast<ARGS_T&>(topData[INDEX_T]) ...);
+    }
+
+    template<typename ... ARGS_T>
+    static TopTaskStatus wrapper([[maybe_unused]] WorkerContext& rCtx, ArrayView<entt::any> topData) noexcept
+    {
+        assert(topData.size() == sizeof...(ARGS_T));
+
+        return apply<ARGS_T ...>(topData, std::make_integer_sequence<std::size_t, sizeof...(ARGS_T)>{});
+    }
+
+    template<typename ... ARGS_T>
+    static constexpr TopTaskFunc_t unpack([[maybe_unused]] TopTaskStatus(*func)(ARGS_T...))
+    {
+        // Parameter pack used to extract function arguments, func is discarded
+        return &wrapper<ARGS_T ...>;
+    }
+
+    template<typename ... ARGS_T>
+    static constexpr TopTaskFunc_t unpack([[maybe_unused]] void(*func)(ARGS_T...))
+    {
+        // Parameter pack used to extract function arguments, func is discarded
+        return &wrapper_void<ARGS_T ...>;
+    }
+};
+
+/**
+ * @brief Wrap a function with arbitrary arguments into a TopTaskFunc_t
+ *
+ * A regular TopTaskFunc_t accepts a ArrayView<entt::any> for passing data of
+ * arbitrary types. This needs to be manually casted.
+ *
+ * wrap_args creates a wrapper function that will automatically cast the
+ * entt::anys and call the underlying function.
+ *
+ * @param a[in]
+ *
+ * @return TopTaskFunc_t function pointer to wrapper
+ */
+template<typename FUNC_T>
+constexpr TopTaskFunc_t wrap_args(FUNC_T const&& funcArg)
+{
+    // TODO: wrap function pointers into a functor
+    static_assert ( ! std::is_function_v<FUNC_T>, "Support for function pointers not yet implemented");
+
+    using static_lambda_t = std::remove_cvref_t<decltype(funcArg)>;
```

Comment:
This seems like a pretty big assumption.

I'm struggling to understand what your code is trying to accomplish.

Reply:
I think this line in particular was left over from some removed code? It still worked when removed, and FUNC_T can be used directly.